pkg/common/cns-lib/volume/manager.go

```diff
@@ -565,63 +576,185 @@ func (m *defaultManager) DeleteVolume(ctx context.Context, volumeID string, dele
 			log.Errorf("ConnectCns failed with err: %+v", err)
 			return err
 		}
-		// Construct the CNS VolumeId list.
+		if m.idempotencyHandlingEnabled {
+			return m.deleteVolumeWithImprovedIdempotency(ctx, volumeID, deleteDisk)
+		}
+		return m.deleteVolume(ctx, volumeID, deleteDisk)
+	}
+	start := time.Now()
+	err := internalDeleteVolume()
+	if err != nil {
```

Comment:
L586-592 and L737-740 have duplicated code. It would be better to put them into one common method to reduce code size.

Reply:
@zhelongp the github diff for this change is confusing, but what its showing is that those lines were removed from L737 and added here. 
We only have a single Prometheus metrics `if-else` for `DeleteVolume`, irrespective of whether the fss is enabled or not.

Reply:
How about other methods that add prometheus metrics in the same file?

You may pass "err" and "metric" as parameters to a common method for prometheus calls.
Then, all callers will call this single method, instead of duplicating the if-else statement.
Your code size will reduce. Source code will be more succinct to read too.

Reply:
Yes i agree, i just saw that you closed https://github.com/kubernetes-sigs/vsphere-csi-driver/pull/905/
I'm also not sure if we need prometheus metrics at both layers - CSI APIs and Volume manager (since CSI calls volume manager).

However i'll take up this comment in a dedicated PR. This one is already quite complicated and the focus is on introducing the idempotency changes to delete volume.

Reply:
Thanks, Raunak. A different PR is fine.

Reply:
>I'm also not sure if we need prometheus metrics at both layers - CSI APIs and Volume manager

Yes, CSI metrics is for end to end time take by CSI and CNS metrics is for the time taken by CNS as seen by CSI(the client). Both are useful to pin point problems in the SaaS world.